beluga/include/beluga/containers/circular_array.hpp

```diff
@@ -0,0 +1,543 @@
+ * and move-assignment of array elements of type T). Additionally, it features optional indexing
+ * reversal (making it a LIFO rather than a FIFO data structure), optional back value extrapolation
+ * for constant read accesses, and optional rollover on write accesses (i.e. overwriting the oldest
+ * value when pushing to an array instance that is full).
+ *
+ * \tparam T Element type.
+ *   It must be [default constructible](https://en.cppreference.com/w/cpp/named_req/DefaultConstructible),
+ *   [move constructible](https://en.cppreference.com/w/cpp/named_req/MoveConstructible), and
+ *   [move assignable](https://en.cppreference.com/w/cpp/named_req/MoveAssignable).
+ * \tparam N Maximum size. It must be a positive integer.
+ * \tparam F Array feature flags, to enable optional functionality. It defaults to none.
+ */
+template <typename T, std::size_t N, CircularArrayFeatureFlags F = CircularArrayFeatureFlags::kNone>
+class CircularArray {
+ public:
+  /// Value type of the array.
+  using value_type = T;
+  /// Size type of the array.
+  using size_type = std::size_t;
+  /// Size difference type of the array.
+  using difference_type = std::ptrdiff_t;
+  /// Value reference type of the array.
+  using reference = value_type&;
+  /// Constant value reference type of the array.
+  using const_reference = const value_type&;
+  /// Value pointer type of the array.
+  using pointer = value_type*;
+  /// Constant value pointer type of the arra.y
+  using const_pointer = const value_type*;
+
+  /// Allocator type of the array (required in range-v3 10.0).
+  using allocator_type = void;
+
+  /// Iterator type of the array.
+  using iterator = IndexingIterator<CircularArray<T, N, F>>;
+  /// Constant iterator type of the array.
+  using const_iterator = IndexingIterator<const CircularArray<T, N, F>>;
+  /// Reverse iterator type of the array.
+  using reverse_iterator = std::reverse_iterator<iterator>;
+  /// Constant reverse iterator type of the array.
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+  /// Default constructor.
+  /**
+   * Array will be functionally empty but storage will be default initialized.
+   */
+  CircularArray() = default;
+
+  /// Constructs array from a pair of iterators.
+  /**
+   * Functionally equivalent to repeated push_back() operations
+   * or push_front() operations in reverse if the layout reversal
+   * feature is enabled. That is, both range and array layouts match.
+   *
+   * \param first Iterator to the beginning of a range.
+   * \param last Sentinel for the end of a range.
+   * \throws std::length_error If the range does not fit in the array
+   * (and the rollover on write feature is not enabled).
+   */
+  template <
+      typename Iterator,
+      typename Sentinel,
+      typename = std::enable_if_t<std::is_same_v<T, typename std::iterator_traits<Iterator>::value_type>>>
+  CircularArray(Iterator first, Sentinel const last) {
```

Comment:
I think clang-tidy is being silly here.

```suggestion
  CircularArray(Iterator first, Sentinel last) {  // NOLINT(readability-non-const-parameter)
```

Reply:
It's funny though, it only shows on the Noetic container.

Reply:
:shrug: